confgenerator/resourcedetector/gce_detector_test.go

```diff
@@ -0,0 +1,165 @@
+package resourcedetector
+
+import (
+	"fmt"
+	"testing"
+
+	gcp_metadata "cloud.google.com/go/compute/metadata"
+)
+
+func TestGettingDetectorWithoutError(t *testing.T) {
+	fakeProvider := FakeProvider{value: "some_value"}
+	fakeBuilder := GCEDetectorBuilder{&fakeProvider}
+	actual, err := fakeBuilder.GetDetector()
+	if err != nil {
+		t.Errorf("should not have error")
+	}
+	if actual == nil {
+		t.Errorf("should not return nil detector when no error")
+	} else if d, ok := actual.(GCEDetector); ok {
+		if d.InstanceID != "some_value" {
+			t.Errorf("detector attribute InstanceID has wrong value")
+		}
+		if d.InterfaceIPv4["some_value"] != "some_value" {
+			t.Errorf("detector attribute InterfaceIPv4 has wrong value")
+		}
+	} else {
+		t.Errorf("should have created GCEDetector")
+	}
+}
+
+func TestErrorGettingDetector(t *testing.T) {
+	fakeProvider := FakeProvider{err: fmt.Errorf("some error")}
+	fakeBuilder := GCEDetectorBuilder{&fakeProvider}
+	actual, err := fakeBuilder.GetDetector()
+	if err == nil {
+		t.Errorf("should have error")
+	}
+	if err.Error() != "some error" {
+		t.Errorf("should have return the correct error message")
+	}
+	if actual != nil {
+		t.Errorf("should return nil detector when having error")
+	}
+}
+
+func TestGettingDetectorOnGCE(t *testing.T) {
```

Comment:
This feels like it belongs in the `integration_test` folder right?


Also you will be able to test a lot more of the fields there since we have the VM information when the integration test sets it up:
https://github.com/GoogleCloudPlatform/ops-agent/blob/e7aa72d2b077a131440bdc999e6b7f36aba548be/integration_test/gce/gce_testing.go#L267-L282

See this as an example of what is used when setting up the environment for an integration test:
https://github.com/GoogleCloudPlatform/ops-agent/blob/e7aa72d2b077a131440bdc999e6b7f36aba548be/integration_test/ops_agent_test.go#L346

Reply:
I don't know what the best way to integration test this. Maybe the easiest way is to have a small go command line runner that hits your interface and dumps the metadata into a json file. The integration test can then read that from the VM and compare it to what it expects. Just a thought, figured its worth sharing. Looking forward to whatever you end up going with.